kubectl replace --grace-period=0 --force ends with nil pointer dereference

<!-- This form is for bug reports and feature requests ONLY!

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).

If this may be security issue, please disclose it privately via https://kubernetes.io/security/.
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug
/sig CLI

**What happened**:
Running `./kubernetes/cluster/kubectl.sh replace -f pod2.yaml --grace-period=0 --force` leads to nil pointer dereference:

```
./kubernetes/cluster/kubectl.sh replace -f pod2.yaml --grace-period=0 --force
pod "nginx" deleted
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1028416]

goroutine 1 [running]:
k8s.io/kubernetes/pkg/kubectl/genericclioptions/resource.(*Result).Visit(0x0, 0xc420fa7910, 0x0, 0x124d21d)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/genericclioptions/resource/result.go:96 +0x26
k8s.io/kubernetes/pkg/kubectl/cmd/wait.(*WaitOptions).RunWait(0xc4206f2150, 0x0, 0x184a9c0)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/wait/wait.go:168 +0x9a
k8s.io/kubernetes/pkg/kubectl/cmd.(*DeleteOptions).DeleteResult(0xc4200bed80, 0xc4200bd000, 0xc4200bed80, 0xc4206c8f20)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/delete.go:286 +0x287
k8s.io/kubernetes/pkg/kubectl/cmd.(*ReplaceOptions).forceReplace(0xc4200be480, 0x0, 0x0)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/replace.go:276 +0x368
k8s.io/kubernetes/pkg/kubectl/cmd.(*ReplaceOptions).Run(0xc4200be480, 0x0, 0x0)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/replace.go:208 +0x1d6
k8s.io/kubernetes/pkg/kubectl/cmd.NewCmdReplace.func1(0xc42054aa00, 0xc42007b7c0, 0x0, 0x4)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/replace.go:119 +0xc3
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc42054aa00, 0xc42007b740, 0x4, 0x4, 0xc42054aa00, 0xc42007b740)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:760 +0x2c1
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420140280, 0xc42079d3e0, 0x12a05f200, 0xc420857ee8)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:846 +0x30a
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420140280, 0x17a38d8, 0x2388d00)
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:794 +0x2b
main.main()
	/home/nilebox/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:50 +0x196
```

**What you expected to happen**:
`kubectl replace` is expected to run delete with wait in that case

**How to reproduce it (as minimally and precisely as possible)**:
`./kubernetes/cluster/kubectl.sh replace -f pod2.yaml --grace-period=0 --force` (with any valid resource yaml)


**Anything else we need to know?**:
Discovered in #64375 when accidentally enabled wait for deletion from other commands than `kubectl delete` and saw failing integration tests in https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64375/pull-kubernetes-integration/13560/

**Environment**:
- Kubernetes version (use `kubectl version`): latest master 
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/pkg/kubectl/cmd/delete.go
+++ b/pkg/kubectl/cmd/delete.go
@@ -139,8 +139,6 @@ func NewCmdDelete(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra
 
 	deleteFlags.AddFlags(cmd)
 
-	cmd.Flags().Bool("wait", true, `If true, wait for resources to be gone before returning.  This waits for finalizers.`)
-
 	cmdutil.AddIncludeUninitializedFlag(cmd)
 	return cmd
 }
@@ -165,13 +163,8 @@ func (o *DeleteOptions) Complete(f cmdutil.Factory, args []string, cmd *cobra.Co
 	}
 	if o.GracePeriod == 0 && !o.ForceDeletion {
 		// To preserve backwards compatibility, but prevent accidental data loss, we convert --grace-period=0
-		// into --grace-period=1 and wait until the object is successfully deleted. Users may provide --force
-		// to bypass this wait.
-		o.WaitForDeletion = true
+		// into --grace-period=1. Users may provide --force to bypass this conversion.
 		o.GracePeriod = 1
-	}
-	if b, err := cmd.Flags().GetBool("wait"); err == nil {
-		o.WaitForDeletion = b
 	}
 
 	includeUninitialized := cmdutil.ShouldIncludeUninitialized(cmd, false)
@@ -216,6 +209,11 @@ func (o *DeleteOptions) Validate(cmd *cobra.Command) error {
 	}
 	if o.DeleteAll && len(o.FieldSelector) > 0 {
 		return fmt.Errorf("cannot set --all and --field-selector at the same time")
+	}
+
+	if o.GracePeriod == 0 && !o.ForceDeletion && !o.WaitForDeletion {
+		// With the explicit --wait flag we need extra validation for backward compatibility
+		return fmt.Errorf("--grace-period=0 must have either --force specified, or --wait to be set to true")
 	}
 
 	switch {

--- a/pkg/kubectl/cmd/delete_flags.go
+++ b/pkg/kubectl/cmd/delete_flags.go
@@ -39,6 +39,7 @@ type DeleteFlags struct {
 	IgnoreNotFound *bool
 	Now            *bool
 	Timeout        *time.Duration
+	Wait           *bool
 	Output         *string
 }
 
@@ -85,6 +86,9 @@ func (f *DeleteFlags) ToOptions(dynamicClient dynamic.Interface, streams generic
 	if f.Timeout != nil {
 		options.Timeout = *f.Timeout
 	}
+	if f.Wait != nil {
+		options.WaitForDeletion = *f.Wait
+	}
 
 	return options
 }
@@ -118,11 +122,12 @@ func (f *DeleteFlags) AddFlags(cmd *cobra.Command) {
 	if f.IgnoreNotFound != nil {
 		cmd.Flags().BoolVar(f.IgnoreNotFound, "ignore-not-found", *f.IgnoreNotFound, "Treat \"resource not found\" as a successful delete. Defaults to \"true\" when --all is specified.")
 	}
-
+	if f.Wait != nil {
+		cmd.Flags().BoolVar(f.Wait, "wait", *f.Wait, "If true, wait for resources to be gone before returning. This waits for finalizers.")
+	}
 	if f.Output != nil {
 		cmd.Flags().StringVarP(f.Output, "output", "o", *f.Output, "Output mode. Use \"-o name\" for shorter output (resource/name).")
 	}
-
 }
 
 // NewDeleteCommandFlags provides default flags and values for use with the "delete" command
@@ -139,6 +144,7 @@ func NewDeleteCommandFlags(usage string) *DeleteFlags {
 	labelSelector := ""
 	fieldSelector := ""
 	timeout := time.Duration(0)
+	wait := true
 
 	filenames := []string{}
 	recursive := false
@@ -156,6 +162,7 @@ func NewDeleteCommandFlags(usage string) *DeleteFlags {
 		IgnoreNotFound: &ignoreNotFound,
 		Now:            &now,
 		Timeout:        &timeout,
+		Wait:           &wait,
 		Output:         &output,
 	}
 }
@@ -167,6 +174,7 @@ func NewDeleteFlags(usage string) *DeleteFlags {
 
 	force := false
 	timeout := time.Duration(0)
+	wait := false
 
 	filenames := []string{}
 	recursive := false
@@ -180,5 +188,6 @@ func NewDeleteFlags(usage string) *DeleteFlags {
 		// add non-defaults
 		Force:   &force,
 		Timeout: &timeout,
+		Wait:    &wait,
 	}
 }

